core/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java

```diff
@@ -268,27 +278,30 @@ private void executeSearch(SearchTask task, SearchTimeProvider timeProvider, Sea
             Collections.unmodifiableMap(aliasFilter), concreteIndexBoosts, listener).start();
     }
 
-    private static GroupShardsIterator mergeShardsIterators(GroupShardsIterator localShardsIterator,
-                                                            List<ShardIterator> remoteShardIterators) {
-        if (remoteShardIterators.isEmpty()) {
-            return localShardsIterator;
-        }
-        List<ShardIterator> shards = new ArrayList<>();
-        for (ShardIterator shardIterator : remoteShardIterators) {
+    private static GroupShardsIterator<SearchShardIterator> mergeShardsIterators(GroupShardsIterator<ShardIterator> localShardsIterator,
+                                                             OriginalIndices locallIndices,
+                                                             List<SearchShardIterator> remoteShardIterators) {
+        List<SearchShardIterator> shards = new ArrayList<>();
+        for (SearchShardIterator shardIterator : remoteShardIterators) {
             shards.add(shardIterator);
         }
         for (ShardIterator shardIterator : localShardsIterator) {
-            shards.add(shardIterator);
+            List<ShardRouting> shardRoutings = new ArrayList<>();
+            for (ShardRouting shardRouting : shardIterator.asUnordered()) {
```

Comment:
I hate this copying here :) can we maybe do this:
 * make `ShardsIterator extends Iterable<ShardRouting>`
 * remove `ShardsIterator#asUnordered()`
 * add `List<ShardRouting> ShardsIterator#getShardRoutings()`

that way we can just reuse the list?